### PR TITLE
[IMM32] Rewrite ImmEnumInputContext

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose
Implementing Japanese input...

JIRA issue: [CORE-11700](https://jira.reactos.org/browse/CORE-11700)

## Proposed changes

- Add `Imm32IsValidHimc`, `Imm32BuildHimcList` and `Imm32AllocAndBuildHimcList` helper functions.
- Rewrite `ImmEnumInputContext` function by using them.
- Modify `NtUserBuildHimcList`.

### CODE REVIEW

File: dll/win32/imm32/imm.c [hunk 1]
```diff
@@ -1161,6 +1161,55 @@ static PCLIENTIMC APIENTRY Imm32GetClientImcCache(void)
     return NULL;
 }
 
+static BOOL Imm32IsValidHimc(HIMC hIMC)
+{
+    return (hIMC != NULL); // FIXME
```
Comment (learn-more) [inline dll/win32/imm32/imm.c:1166]: If you give imm32 it's own heap for `Imm32HeapAlloc`, and you allocate an HIMC from this heap,
you can use `RtlValidateHeap` to check that it was allocated from the correct heap.
That seems like a good starting point to validate an HIMC?

Comment (katahiromz) [reply dll/win32/imm32/imm.c]: OK. Thanks.

File: dll/win32/imm32/imm.c [hunk 2]
```diff
@@ -3735,11 +3784,31 @@ BOOL WINAPI ImmDisableTextFrameService(DWORD dwThreadId)
...
+    TRACE("ImmEnumInputContext(%lu, %p, %p)\n", dwThreadId, lpfn, lParam);
+
+    dwCount = Imm32AllocAndBuildHimcList(dwThreadId, &phList);
+    if (!dwCount)
+        return FALSE;
+
+    for (dwIndex = 0; dwIndex < dwCount; ++dwIndex)
+    {
+        hIMC = phList[dwIndex];
+        if (!Imm32IsValidHimc(hIMC))
```
Comment (HBelusca) [inline dll/win32/imm32/imm.c:3803]: Why wouldn't it be valid? that hIMC entry comes directly from that allocated phList by Imm32AllocAndBuildHimcList and initialized by Imm32BuildHimcList(), and so its contents must be valid by construction ?

Comment (katahiromz) [reply dll/win32/imm32/imm.c]: I deleted the check.

**Pull request merged**